Add support for light color modes

emontnemery opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add support for light color modes, background in https://github.com/home-assistant/architecture/discussions/519.
The color mode will be stored in the light's state, but should *not* be passed to the `turn_on` service call.
When turning on the light, the color mode is non ambiguous because only one color may be given.

Colors will no longer be force-converted through hs color space when updating state or in a `turn_on` call, instead the light's native color space will be respected.

New attributes:
```
ATTR_COLOR_MODE
ATTR_SUPPORTED_COLOR_MODES
ATTR_RGBW_COLOR
ATTR_RGBWW_COLOR
```

Defined color modes:
```
COLOR_MODES = {
    COLOR_MODE_ONOFF,  # The light can be turned on or off. This mode must be the only supported mode if supported by the light
    COLOR_MODE_DIMMER,  # The light can be dimmed. This mode must be the only supported mode if supported by the light
    COLOR_MODE_COLOR_TEMP,  # The light can be dimmed and its color temperature is present in the state.
    COLOR_MODE_HS,  # The light can be dimmed and its color is present in the state as an (h, s) tuple (no brightness).
    COLOR_MODE_XY,  # The light can be dimmed and its color is present in the state as an (x, y) tuple (no brightness).
    COLOR_MODE_RGB,  # The light can be dimmed and its color is present in the state as an (r, g, b) tuple (not normalized).
    COLOR_MODE_RGBW,  # The light can be dimmed and its color is present in the state as an (r, g, b, w) tuple (not normalized).
    COLOR_MODE_RGBWW,  # The light can be dimmed and its color is present in the state as an (r, g, b, cw, ww) tuple (not normalized).
    COLOR_MODE_UNKNOWN,  # The light has reported an ambiguous state (e.g. no brightness for a light which supports `COLOR_MODE_DIMMER`)
}
```

New methods added to the base component `LightEntity`:
- `color_mode`
- `supported_color_modes`
- `rgb_color`
- `rgbw_color`
- `rgbww_color`
- `xy_color`

For backwards compatibility:
- If the light does not implement `supported_color_modes`, it will be calculeted based on `supported_features`
- If the light does not implement `color_mode`, current `color_mode` will be guessed in the following order:
   - `COLOR_MODE_RGBW` if the light's `white_value` and `hs_color` are not `None`
   - `COLOR_MODE_HS` if the light's `hs_color` is not `None`
   - `COLOR_MODE_COLOR_TEMP` if the light's `color_temp` is not None
   - `COLOR_MODE_DIMMER` if the light's `brihtness` is not None
   - `COLOR_MODE_NONE` otherwise
- If the light does not implement `supported_color_modes`, a `turn_on` service call with `ATTR_RGBW_COLOR` will be converted to `ATTR_HS_COLOR` + `ATTR_WHITE_VALUE`.

Deprecated:
```
SUPPORT_BRIGHTNESS = 1
SUPPORT_COLOR_TEMP = 2
SUPPORT_COLOR = 16
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

emontnemery committed:
Add support for light color modes

emontnemery committed:
Update tests

emontnemery committed:
Update comments

balloob commented on homeassistant/components/light/__init__.py:
```diff
@@ -225,6 +252,41 @@ async def async_handle_light_on_service(light, call):
+        # If a color is specified, convert to the color space supported by the light
+        # Backwards compatibility: Fall back to hs color if light.supported_color_modes
+        # is not implemented
+        if ATTR_HS_COLOR or ATTR_RGB_COLOR or ATTR_XY_COLOR in params:
```
This tests if `ATTR_HS_COLOR` is truthy, `ATTR_RGB_COLOR` is truthy, or `ATTR_XY_COLOR in params`. The first two are defined so this code is always executed?
```suggestion
        if ATTR_HS_COLOR in params or ATTR_RGB_COLOR in params or ATTR_XY_COLOR in params:
```

emontnemery replied:
> I would expect Google, Alexa and Homekit requiring updates because of this PR

Does it mean those integrations will need to be updated to handle the new colour mode, rgbw and rgbww?

emontnemery replied:
Oops, tests are needed for the new code

Kane610 replied:
Couldn't you use intersection here?

balloob replied:
Not sure if more readable:

```python
if params.keys() & {ATTR_HS_COLOR, ATTR_RGB_COLOR, ATTR_XY_COLOR}:
```

You can also do

```python
if any(key in params for key in (ATTR_HS_COLOR, ATTR_RGB_COLOR, ATTR_XY_COLOR)):
```

balloob commented on homeassistant/components/light/__init__.py:
```diff
@@ -225,6 +252,41 @@ async def async_handle_light_on_service(light, call):
+                if (xy_color := params.pop(ATTR_XY_COLOR, None)) is not None:
+                    params[ATTR_HS_COLOR] = color_util.color_xy_to_hs(*xy_color)
+
+            elif COLOR_MODE_RGB in supported_color_modes:
```
Make sure every fallback has a test.

emontnemery replied:
Sure, the tests for the light component needs to be updated to test the new features as well as all the backwards compatibility cases.

balloob commented on homeassistant/components/light/__init__.py:
```diff
@@ -411,11 +473,79 @@ def brightness(self) -> Optional[int]:
+        ):
+            # Backwards compatibility for rgbw_color added in 2021.4
+            # Add warning in 2021.6, remove in 2021.10
+            r, g, b = color_util.color_hs_to_RGB(*self.hs_color)
```
```suggestion
            # pylint: disable=invalid-name
            r, g, b = color_util.color_hs_to_RGB(*self.hs_color)
```

emontnemery commented:
> I would expect Google, Alexa and Homekit requiring updates because of this PR

Does it mean those integrations will need to be updated to handle the new colour modes, rgbw and rgbww?

emontnemery committed:
Fix bugs, add tests

emontnemery committed:
Suppress lint errors

emontnemery committed:
Don't suppress brightness when state is ambiguous

emontnemery commented:
@koenkk could you have a look at this from a z2m perspective?

emontnemery committed:
Improve reproduce_state + add tests

emontnemery commented:
@zha, @Kane610 Can you review the proposal from the perspective of zha / deconz?

zha commented:
> @zha, @Kane610 Can you review the proposal from the perspective of zha / deconz?

I am sure you mentioned the wrong person 😊

Adminiuga commented:
Nothing jumps out for ZHA. LGTM

Please notify me when this gets merged (in case I miss it), and I'll test it, update ZHA if needed.

Koenkk commented:
@emontnemery looks good, 2 comments however:
- `COLOR_MODE_BRIGHTNESS` looks like a more logical name for `COLOR_MODE_DIMMER`
- Based on this: https://github.com/home-assistant/core/pull/47720/files#diff-1d3290e4b1b6fbd972cf52246c5138aef77c3dc2419099576d5d2e0f7cc842cdR433 I expect the MQTT payload should contain the `color_mode` like this (lets say we have a bulb which is currently in XY mode):
```js
{"state": "ON", "brightness": 100, "color": {"x": 0.7, "y": 0.3"}, "color_mode": "color_mode_xy"}
```

The `"color_mode": "color_mode_xy"` looks a bit strange to me since `color_mode` is in both the property and value. What about just: `"color_mode": "xy"`?

balloob commented on homeassistant/components/light/__init__.py:
```diff
@@ -228,6 +256,48 @@ async def async_handle_light_on_service(light, call):
+                params[ATTR_HS_COLOR] = color_util.color_RGB_to_hs(*rgb_color)
+            elif (xy_color := params.pop(ATTR_XY_COLOR, None)) is not None:
+                params[ATTR_HS_COLOR] = color_util.color_xy_to_hs(*xy_color)
+        elif ATTR_HS_COLOR in params and COLOR_MODE_HS not in supported_color_modes:
+            hs_color = params.pop(ATTR_HS_COLOR)
```
You could go all wallrus:

```suggestion
        elif COLOR_MODE_HS not in supported_color_modes and (hs_color := params.pop(ATTR_HS_COLOR, None) is not None:
```

emontnemery replied:
It becomes 4 lines instead 2 after formatting though:
```py
elif (
            COLOR_MODE_HS not in supported_color_modes
            and (hs_color := params.pop(ATTR_HS_COLOR, None)) is not None
        ):
```

balloob commented on homeassistant/components/light/__init__.py:
```diff
@@ -473,27 +638,84 @@ def state_attributes(self):
+        color_mode = self._light_internal_color_mode
+
+        if color_mode not in self._light_internal_supported_color_modes:
+            _LOGGER.debug(
```
Should we log this as `error` the first time per-entity?

emontnemery replied:
I think we should start doing that in 1-2 releases, not yet. I added a comment.

emontnemery committed:
Add comment

emontnemery committed:
Change COLOR_MODE_* constants, rename COLOR_MODE_DIMMER to COLOR_MODE_BRIGHTNESS

emontnemery commented:
@Koenkk 
 
> * `COLOR_MODE_BRIGHTNESS` looks like a more logical name for `COLOR_MODE_DIMMER`

I agree, changed.

> * Question: what happens if there is no `color_mode` property in the payload? I would expect `COLOR_MODE_UNKNOWN` to be used in this case.

For an MQTT JSON light, I think we should add two new configuration variables:
 - `color_mode`: set to `True` if the light will report its color mode, `False` otherwise. 
 - `supported_color_modes`: A list of supported color modes
For a light which has is configured with `color_mode=True`, it would be an error if we receive a state update dict without the `color_mode`, or a `color_mode` not included in `supported_color_modes`.
 
> The `"color_mode": "color_mode_xy"` looks a bit strange to me since `color_mode` is in both the property and value. What about just: `"color_mode": "xy"`?

I agree, changed.

emontnemery committed:
Fix tests

emontnemery committed:
Tweaks

Koenkk commented:
@emontnemery big thanks, when looking at the changes files I don't see anything changed in MQTT. Does this mean a new merge request has to be done for this? (or can I already test this out somehow)
